odoo/tools/translate.py

```diff
@@ -476,251 +477,288 @@ def unquote(str):
+        return self.reader.__iter__()
+
+class CSVFileReader:
+    def __init__(self, source):
+        self.source = pycompat.csv_reader(source, quotechar='"', delimiter=',')
+        # read the first line of the file (it contains columns titles)
+        self.fields = next(self.source)
 
     def __iter__(self):
-        self.buffer.seek(0)
-        self.lines = self._get_lines()
-        self.lines_count = len(self.lines)
-
-        self.first = True
-        self.extra_lines= []
-        return self
-
-    def _get_lines(self):
-        lines = self.buffer.readlines()
-        # remove the BOM (Byte Order Mark):
-        if len(lines):
-            lines[0] = lines[0].lstrip(u"\ufeff")
-
-        lines.append('') # ensure that the file ends with at least an empty line
-        return lines
-
-    def cur_line(self):
-        return self.lines_count - len(self.lines)
-
-    def next(self):
-        trans_type = name = res_id = source = trad = None
-        if self.extra_lines:
-            trans_type, name, res_id, source, trad, comments = self.extra_lines.pop(0)
-            if not res_id:
-                res_id = '0'
+        for entry in self.source:
+            yield zip(self.fields, entry)
+
+class PoFileReader:
+    """ Iterate over po file to return Odoo translation entries """
+    def __init__(self, source):
+
+        def get_pot_path(source_name):
+            # when fileobj is a TemporaryFile, its name is an interget in P3, a string in P2
+            if isinstance(source_name, str) and source_name.endswith('.po'):
+                # Normally the path looks like /path/to/xxx/i18n/lang.po
+                # and we try to find the corresponding
+                # /path/to/xxx/i18n/xxx.pot file.
+                # (Sometimes we have 'i18n_extra' instead of just 'i18n')
+                addons_module_i18n, _ignored = os.path.split(source_name)
+                addons_module, i18n_dir = os.path.split(addons_module_i18n)
+                addons, module = os.path.split(addons_module)
+                return file_open(os.path.join(
+                    addons, module, i18n_dir, module + '.pot'), mode='rb')
+            return False
+
+        # polib accepts either a path, either the file content as a string
+        # NamedTemporaryFile is not a BufferedIOBase but a wrapper
+        if isinstance(source, io.BufferedIOBase) or \
+            (hasattr(source, 'file') and isinstance(source.file, io.BufferedIOBase)):
+            self.pofile = polib.pofile(source.read().decode())
+            pot_path = get_pot_path(source.name)
+
+        else:
+            self.pofile = polib.pofile(source)
+            pot_path = get_pot_path(source)
+
+        if pot_path:
+            # Make a reader for the POT file
+            # (Because the POT comments are correct on GitHub but the
+            # PO comments tends to be outdated. See LP bug 933496.)
+            pot_reader = PoFileReader(pot_path)
+            self.pofile.merge(pot_reader.pofile)
```

Comment:
Using `PoFileReader` recursively is confusing here.
It seems that you may end up looking up recursively for a `pot` file...

Why change `get_pot_path` returns a pathname (a string or `None`), and then simply:
```suggestion
            self.pofile.merge(polib.pofile(pot_path))
```